File: objectiveai/objectiveai/src/functions/executions/response/unary/function_execution.rs
use crate::{
    error,
    functions::{self, executions::response},
    vector,
};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FunctionExecution {
    pub id: String,
    pub tasks: Vec<super::Task>,
    pub tasks_errors: bool,
    pub reasoning: Option<super::ReasoningSummary>,
    pub output: functions::expression::FunctionOutput,
    pub error: Option<error::ResponseError>,
    pub retry_token: Option<String>,
    pub created: u64,
    pub function: Option<String>,
    pub profile: Option<String>,
    pub object: super::Object,
    pub usage: vector::completions::response::Usage,
}

impl FunctionExecution {
    pub fn any_usage(&self) -> bool {
        self.usage.any_usage()
    }
}

impl From<response::streaming::FunctionExecutionChunk> for FunctionExecution {
    fn from(
        response::streaming::FunctionExecutionChunk {
            id,
            tasks,
            tasks_errors,
            reasoning,
            output,
            error,
            retry_token,
            created,
            function,
            profile,
            object,
            usage,
        }: response::streaming::FunctionExecutionChunk,
    ) -> Self {
        Self {
            id,
            tasks: tasks.into_iter().map(super::Task::from).collect(),
            tasks_errors: tasks_errors.unwrap_or(false),
            reasoning: reasoning.map(super::ReasoningSummary::from),
            output: output.unwrap_or(
                functions::expression::FunctionOutput::Err(
                    serde_json::Value::Null,
                ),
            ),
            error,
            retry_token,
            created,
            function,
            profile,
            object: object.into(),
            usage: usage.unwrap_or_default(),
        }
    }
}

File: objectiveai-js/src/functions/profiles/computations/response/unary/functionExecution.ts
import { z } from "zod";
import { AgentCompletionsResponseUsageSchema } from "../../../../../agent/completions/response/usage";
import { FunctionsExecutionsResponseUnaryObjectSchema } from "../../../../executions/response/unary/object";
import { FunctionsExecutionsResponseUnaryReasoningSummarySchema } from "../../../../executions/response/unary/reasoningSummary";
import { FunctionsExecutionsResponseUnaryTaskSchema } from "../../../../executions/response/unary/task";
import { FunctionsExpressionTaskOutputOwnedSchema } from "../../../../expression/taskOutputOwned";
import { ResponseErrorSchema } from "../../../../../responseError";

export const FunctionsProfilesComputationsResponseUnaryFunctionExecutionSchema = z.object({
  index: z.number().int().min(0).meta({ format: "uint64" }),
  dataset: z.number().int().min(0).meta({ format: "uint64" }),
  n: z.number().int().min(0).meta({ format: "uint64" }),
  retry: z.number().int().min(0).meta({ format: "uint64" }),
  id: z.string().describe("Unique identifier for this execution."),
  tasks: z.array(FunctionsExecutionsResponseUnaryTaskSchema).describe("Results from each task in the function."),
  tasks_errors: z.boolean().describe("Whether any tasks encountered errors."),
  reasoning: FunctionsExecutionsResponseUnaryReasoningSummarySchema.nullable().describe("Reasoning summary if reasoning was enabled.").optional(),
  output: FunctionsExpressionTaskOutputOwnedSchema.describe("The final output (scalar or vector score)."),
  error: ResponseErrorSchema.nullable().describe("Error details if the execution failed.").optional(),
  retry_token: z.string().nullable().describe("Token for retrying this execution with cached votes.").optional(),
  created: z.number().int().min(0).meta({ format: "uint64" }).describe("Unix timestamp when the execution was created."),
  function: z.string().nullable().describe("ID of the function used (if remote).").optional(),
  profile: z.string().nullable().describe("ID of the profile used (if remote).").optional(),
  object: FunctionsExecutionsResponseUnaryObjectSchema.describe("Object type identifier."),
  usage: AgentCompletionsResponseUsageSchema.describe("Aggregated token and cost usage."),
}).describe("A complete function execution response (non-streaming).").meta({ title: "functions.profiles.computations.response.unary.FunctionExecution" });
export type FunctionsProfilesComputationsResponseUnaryFunctionExecution = z.infer<typeof FunctionsProfilesComputationsResponseUnaryFunctionExecutionSchema>;
